### PR TITLE
enhancement : recursive uninstallation of dependent packages

### PR DESCRIPTION
**Pull request by alalazo**
##### Modifications : 
- [x] added `--dependents` option to uninstall
- [x] always prompt user for confirmation
- [x] add tests for regression on the command
- [x] updated documentation

##### Notes : 
This is meant to address #651 . Any comment or bug report on the PR is welcome.

##### Example : 
```
$ spack find
==> 8 installed packages.
-- linux-x86_64 / gcc@4.8 ---------------------------------------
bzip2@1.0.6  ncurses@6.0  openssl@1.0.2g  python@2.7.11  python@3.5.1  readline@6.3  sqlite@3.8.5  zlib@1.2.8

$ spack uninstall python
==> Error: python matches multiple packages:

-- linux-x86_64 / gcc@4.8 ---------------------------------------
nk2wnn4 python@2.7.11  7hxo5t5 python@3.5.1

==> Error: You can either:
    a) Use a more specific spec, or
    b) use spack uninstall -a to uninstall ALL matching specs.

$ spack uninstall -a python zlib
==> Error: Will not uninstall zlib@1.2.8%gcc@4.8-pdtswcf

The following packages depend on it:
-- linux-x86_64 / gcc@4.8 ---------------------------------------
3trgywi openssl@1.0.2g

==> Error: You can use spack uninstall -f to force this action

$ spack uninstall -a -d python zlib
==> The following packages will be uninstalled : 

-- linux-x86_64 / gcc@4.8 ---------------------------------------
3trgywi openssl@1.0.2g  nk2wnn4 python@2.7.11  7hxo5t5 python@3.5.1  pdtswcf zlib@1.2.8

==> Do you want to proceed ? [y/n]
y
==> Successfully uninstalled python@2.7.11%gcc@4.8=linux-x86_64-nk2wnn4
==> Successfully uninstalled python@3.5.1%gcc@4.8=linux-x86_64-7hxo5t5
==> Successfully uninstalled openssl@1.0.2g%gcc@4.8=linux-x86_64-3trgywi
==> Successfully uninstalled zlib@1.2.8%gcc@4.8=linux-x86_64-pdtswcf
```

### CODE REVIEW

File: lib/spack/spack/cmd/uninstall.py [hunk 1]
```diff
@@ -44,10 +49,81 @@ def setup_parser(subparser):
...
         '-a', '--all', action='store_true', dest='all',
         help="USE CAREFULLY. Remove ALL installed packages that match each " +
-        "supplied spec. i.e., if you say uninstall libelf, ALL versions of " +
-        "libelf are uninstalled. This is both useful and dangerous, like rm -r.")
+             "supplied spec. i.e., if you say uninstall libelf, ALL versions of " +
+             "libelf are uninstalled. This is both useful and dangerous, like rm -r.")
     subparser.add_argument(
-        'packages', nargs=argparse.REMAINDER, help="specs of packages to uninstall")
+        '-r', '--recursive', action='store_true', dest='recursive',
+        help='Uninstall all the packages that depends on the ones for which we required explicit removal.'
```
Comment (adamjstewart) [inline lib/spack/spack/cmd/uninstall.py]: I would reword this as "Also uninstall any packages that depend on this package."

**Comment by alalazo** [comment]
@tgamblin Is there already a unit test on a command? I would like to add some unit tests to `spack uninstall` and was wondering if I could take something else as an example.

**Comment by adamjstewart** [comment]
```
$ spack uninstall -a python zlib
==> Error: Will not uninstall zlib@1.2.8%gcc@4.8-pdtswcf

The following packages depend on it:
-- linux-x86_64 / gcc@4.8 ---------------------------------------
3trgywi openssl@1.0.2g

==> Error: You can use spack uninstall -f to force this action
```
Can you add:
```
==> Error: You can use spack uninstall -r to uninstall these dependencies as well"
```
or something like that?

**Comment by citibeth** [comment]
There should be an option to NOT prompt for confirmation.  This will be needed in unit tests, scripts, etc.

**Comment by adamjstewart** [comment]
> There should be an option to NOT prompt for confirmation. This will be needed in unit tests, scripts, etc.

Agreed. How about `spack uninstall -rf`?

**Comment by citibeth** [comment]
I like that, it mirrors rm -rf.

**Comment by alalazo** [comment]
@adamjstewart @citibeth If you look at the PR there is an option to do that:
```
subparser.add_argument(
    '-y', '--yes-to-all', action='store_true', dest='yes_to_all',
    help='Assume "yes" is the answer to every confirmation asked to the user.'
)
```
:smile:

**Comment by alalazo** [comment]
@tgamblin @citibeth @adamjstewart : the PR is for me ready to be merged.

**Comment by trws** [comment]
I like this concept, and I realize it's late to be nitpicking, but is this what people will expect when they see "recursive?"  This is uninstalling all reverse dependencies on a given package.  At least for me, I would kinda expect recursive to be uninstalling a package and everything it caused to be installed that is not otherwise required by an explicitly installed package.

**Comment by citibeth** [comment]
I thought about this too.  Recursing in both directions is something
someone might want.  I think this PR should get merged; and then if someone
wants the other kind of recursion, they can put it in later.

The only thing we MIGHT want to think about now is how the two kinds of
recursion should be named.  Even so... if that results in a UI change, I
think we should merge this one now, and then change the UI in another PR.

**Comment by trws** [comment]
I'm perfectly good with that, it just struck me that the implementation was different than what I expected from the flag name.  It's no less useful for it, and it's definitely something I would use either way.


________________________________

**Comment by citibeth** [comment]
> the implementation was different than what I expected from the flag name

Yes, that does have me concerned.  How about:

# Remove all the things that X depends on (if reference count is 0)
spack uninstall --remove-dependencies X

# Remove all the things that depend on X
spack uninstall --remove-dependents X

Therefore, the change to make to this PR would be to change '--recursive'
to '--remove-dependents'.  Should be easy, if we can all agree on it.

**Comment by citibeth** [comment]
Can you please add some documentation to the .rst files explaining this new feature?

**Comment by citibeth** [comment]
We also need a recursive deactivate.  See the current state of affairs:
```
$ spack deactivate py-numpy^atlas
==> Error: Cannot deactivate py-numpy@1.10.4%gcc@4.9.3+blas+lapack=linux-x86_64-xbgoiqp beacuse py-netcdf@1.2.3.1%gcc@4.9.3=linux-x86_64-v6ectyw is activated and depends on it.
```
Is this something that would be pretty easy, now that your head is into `spack uninstall --recursive`?

**Comment by alalazo** [comment]
@trws 
> At least for me, I would kinda expect recursive to be uninstalling a package and everything it caused to be installed that is not otherwise required by an explicitly installed package.

To do that we need to know if a package was **explicitly** installed and I think we don't have this information available right now. Am I right @tgamblin ?

@citibeth I'll add a few lines of documentation and I'll try to submit asap another PR for deactivate

**Comment by alalazo** [comment]
@trws @citibeth @tgamblin Renaming the option will require just a few changes : names that fit better than `recursive` are welcome :smile:

**Comment by citibeth** [comment]
> names that fit better than recursive are welcome :smile:

I'm going to suggest we standardize on:
    --dependencies (-d) and --dependents (-D)
Dependents is upper-case because it's following the graph BACKWARDS from how they're specified in the Spack packages.  But really... to avoid confusion, I'll probably just use the long versions.

I'm working on another recursive feature, which will be turned on with the --dependencies option.  (See upcoming PR).

**Comment by citibeth** [comment]
Hmm... I don't like '-d' and '-D'.  How about:
     --dependencies        (-r)
     --dependents            (-R)

**Comment by citibeth** [comment]
See PR #721 for where I used recursive flags, consistent with the proposal here.

**Comment by adamjstewart** [comment]
Being able to "uninstall a package and everything it caused to be installed that is not otherwise required by an explicitly installed package" would be really awesome too. @alalazo: If we don't have this capability right now then don't worry about it. Just something to add to the list of feature requests.

I like `--dependencies` and `--dependents`. Either `-r` or `-d` is fine by me.

**Comment by alalazo** [comment]
Done and ready to merge.

**Comment by tgamblin** [comment]
Thanks!
